### DISCUSSION TITLE
Trouble connecting my Node.JS app to the server on Thunder Client.

### DISCUSSION THREAD

**Discussion by remoteconn-7891** [community/community#75567] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

I asked this specific question a few days ago, with no clear solutions. So I'm trying to send a connection to the server for my Node.JS app using Thunder Client on my MacOS. I've tried several troubleshooting methods, including replacing the <localhost> with my MacOS IP Address, '0.0.0.0.' and even relaunched VS Code. My PORT is 3000. It's gotta be something with my PORT. Here is the copy/pasted code from my index.JS entry point and I will also send a screenshot of the error I'm seeing. Please, I've been Googling the answer using various resources with no luck for days.

`const express = require("express");
const cors = require("cors");
const mongoose = require("mongoose");
const dotenv = require("dotenv"); // Import dotenv 
const authRouter = require("./routes/auth");
dotenv.config(); // Load environment variables // Load environment variables from .env file
const PORT = process.env.PORT || 3000;
const app = express();

server.use(express.json());
server.use(authRouter);
//DB just stands for MongoDB database, this String is the password for my database
const DB = 
"mongodb+srv://coreyjames689:remotify87@cluster0.jybj21u.mongodb.net/Cluster0?retryWrites=true&w=majority";

mongoose 
  .connect(DB)
  .then(() => {
    console.log("Connection Successful");
  })
  .catch((e) => {
    console.log(e);
  });
<img width="1440" alt="Screenshot 2023-11-14 at 2 25 23 PM" src="https://github.com/community/community/assets/142646120/e2c76f11-7697-406e-ae27-0182ad6b1c89">

**Comment by stolenfallen1** [comment]
Hi can you try changing this line
```
server.use(express.json());
server.use(authRouter);
```
To this line, you seem to be referencing "server" instead of app.
```
app.use(express.json());
app.use(authRouter);
```
And your code seems to be missing this block of code also:
```
// This code makes sure your app is listening on the specified port you want it to in this case port 3000
app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});
```

**Comment by remoteconn-7891** [comment]
I do apologize for that confusion. I typed 'server' instead of 'app' simply because the guy in the instructional video typed 'server'. I was just following the tutorial video. I'll try this though, thank you

**Comment by remoteconn-7891** [comment]
PS does it matter if I use 'localhost' or my IP Address?

**Comment by remoteconn-7891** [comment]
Nope still not working, even after using the code you suggested. I don't think this is a Firewall issue, as it's disabled

**Reply by stolenfallen1** [reply]
Hmmmm looking back at your screenshot looks like you didn't execute / start the code. 

 In your package.json file, you can see this line of code:
 ```
 "scripts": {
  "start": "node app.js"
}
```

Try running the app by 
```
node app.js
```
or 
```
npm start 
```

**Reply by stolenfallen1** [reply]
And another thing that I have noticed based on your screenshot is that in your Thunder Client, your JSON data's that you are trying to register are wrong. You should do it like this instead:
```
{
    "name:": "your-name" ,
    "email": "your-email",
    "password": "your-password"
}
```

**Reply by remoteconn-7891** [reply]
Ah I see what you did, I will correct the JSON formatting. Here is a copy/paste of my package.JSON:

`{
  "name": "Virtuallify", // My project's name
  "version": "1.0.0", // version of my project
  "description": "Marketplace & professional networking mobile app specifically
  designed for remote work that operates as a B2B SaaS for remote employers", //
  Description of my project
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "Corey James",
  "license": "ISC", // License of my project, still pending
  "dependencies": {
    "bcryptjs": "^2.4.3",
    "express": "^4.18.2",
    "jsonwebtoken": "^9.0.2",
    "mongoose": "^8.0.0"
  },
  "devDependencies": {
    "nodemon": "^3.0.1"
  }
}`

**Reply by remoteconn-7891** [reply]
Sorry for the sloppy code format, this is not how I copied the file in VS Code

**Reply by stolenfallen1** [reply]
It's fine, all my responses are solely base on the information you have given me and the screenshot. 

Try replacing your package.json with this one:
```
{
    "name": "Virtuallify",
    "version": "1.0.0",
    "description": "Marketplace & professional networking mobile app specifically designed for remote work that operates as a B2B SaaS for remote employers",
    "main": "index.js",
    "scripts": {
        // Try adding this line 
        "start": "node index.js",
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "keywords": [],
    "author": "Corey James",
    "license": "ISC",
    "dependencies": {
        "bcryptjs": "^2.4.3",
        "express": "^4.18.2",
        "jsonwebtoken": "^9.0.2",
        "mongoose": "^8.0.0"
    },
    "devDependencies": {
        "nodemon": "^3.0.1"
    }
}
```
Then try re launching your application. 
```
npm start 
```

**Reply by stolenfallen1** [reply]
And last thing I have noticed in your screenshot is that you want to register data that means you need to use POST request in your thunder client but instead you are using GET request. 

Hope it helps :)

**Comment by remoteconn-7891** [comment]
So these are the errors I got when I ran npm start

coreyvigil@Coreys-Air virtualify % npm start
npm ERR! code EJSONPARSE
npm ERR! JSON.parse Invalid package.json: JSONParseError: Bad control character in string literal in JSON at position 123 (line 4 column 80) while parsing '{
npm ERR! JSON.parse "name": "Virtualify",
npm ERR! JSON.parse "version": "2.0.'
npm ERR! JSON.parse Failed to parse JSON data.
npm ERR! JSON.parse Note: package.json must be actual JSON, not just JavaScript.

npm ERR! A complete log of this run can be found in: /Users/coreyvigil/.npm/_logs/2023-11-15T06_45_18_094Z-debug-0.log
coreyvigil@Coreys-Air virtualify %

**Comment by remoteconn-7891** [comment]
Oh I see now, I have to correct this string issue first

**Reply by stolenfallen1** [reply]
Okay go ahead, good luck and happy coding! By the way you can close this GitHub discussion by setting it as answered. :)

**Comment by remoteconn-7891** [comment]
Actually it isn't. Still trying to figure out the string literal error. I think it was a point break. Also in that JSON format in TC, it still shows errors. I'll send you a screenshot of the error and the code I tried to fix in package.JSON.

`{
  "name": "Virtuallify", // My project's name
  "version": "1.0.0", // version of my project
  "description": "Marketplace & professional networking mobile app", // Description of my project
  "main": "index.js",
  "scripts": {
    "start": "node app.js",
    "test": "echo \"Error no test specified\" && exit  1"
  },
  "keywords": [],
  "author": "Corey James",
  "license": "ISC", // License of my project, still pending
  "dependencies": {
    "bcryptjs": "^2.4.3",
    "express": "^4.18.2",
    "jsonwebtoken": "^9.0.2",
    "mongoose": "^8.0.0"
  },
  "devDependencies": {
    "nodemon": "^3.0.1"
  }
}
<img width="1440" alt="Screenshot 2023-11-15 at 12 00 39 AM" src="https://github.com/community/community/assets/142646120/3ed89f5f-87d0-4391-846a-8a6777161c5f">
`

**Reply by stolenfallen1** [reply]
```
{
    "name": "Virtuallify",
    "version": "1.0.0",
    "description": "Marketplace & professional networking mobile app",
    "main": "index.js",
    "scripts": {
        "start": "node app.js",  // And your previous code does not have a space here
        "test": "echo \"Error no test specified\" && exit 1"
    },
    "keywords": [],
    "author": "Corey James",
    "license": "ISC", 
    "dependencies": {
        "bcryptjs": "^2.4.3",
        "express": "^4.18.2",
        "jsonwebtoken": "^9.0.2",
        "mongoose": "^8.0.0"
    },
    "devDependencies": {
        "nodemon": "^3.0.1"
    }
}
// REMOVE THIS Html img element
<img
  width="1440"
  alt="Screenshot 2023-11-15 at 12 00 39 AM"
  src="https://github.com/community/community/assets/142646120/3ed89f5f-87d0-4391-846a-8a6777161c5f"
/>
```

**Comment by remoteconn-7891** [comment]
<img width="1440" alt="Screenshot 2023-11-15 at 12 00 39 AM" src="https://github.com/community/community/assets/142646120/465b1857-4037-44a7-9fdf-8686cdb0846c">

**Reply by stolenfallen1** [reply]
Basing on this screenshot your name does not have a "," comma
```
{
     "name": "example",
     "email": "example",
     "password": "example"   // don't again comma here
}
```

**Reply by remoteconn-7891** [reply]
Okay, that problem is fixed, thank you so much. Now for package.JSON. I'm still trying to figure where that error is coming from, it says line 4. But I think I corrected it

**Comment by remoteconn-7891** [comment]
Sorry about that, was trying to upload my screenshot on here. I'll send it again without that

**Comment by remoteconn-7891** [comment]
`{
  "name": "Virtuallify", // My project's name
  "version": "1.0.0", // version of my project
  "description": "Marketplace & professional networking mobile app", // Description of my project
  "main": "index.js",
  "scripts": {
    "start": "node app.js",
    "test": "echo \"Error no test specified\" && exit  1"
  },
  "keywords": [],
  "author": "Corey James",
  "license": "ISC", // License of my project, still pending
  "dependencies": {
    "bcryptjs": "^2.4.3",
    "express": "^4.18.2",
    "jsonwebtoken": "^9.0.2",
    "mongoose": "^8.0.0"
  },
  "devDependencies": {
    "nodemon": "^3.0.1"
  }
}`

**Reply by stolenfallen1** [reply]
Try copying this code:
```
{
    "name": "Virtuallify",
    "version": "1.0.0",
    "description": "Marketplace & professional networking mobile app",
    "main": "index.js",
    "scripts": {
        "start": "node app.js",
        "test": "echo \"Error no test specified\" && exit  1"
    },
    "keywords": [],
    "author": "Corey James",
    "license": "ISC",
    "dependencies": {
        "bcryptjs": "^2.4.3",
        "express": "^4.18.2",
        "jsonwebtoken": "^9.0.2",
        "mongoose": "^8.0.0"
    },
    "devDependencies": {
        "nodemon": "^3.0.1"
    }
}
```
Afterwards go to VSCode terminal then run
```
npm install
```
Then please try running the app once again.
```
npm start 
```

**Reply by remoteconn-7891** [reply]
But I see what you did, take out the comments. Still doesn't work

**Comment by remoteconn-7891** [comment]
In this error message for npm start, it says the version is "2.0" which conflicts with the version I have in package.jSON which is "1.0.0". Not sure if this is what is causing the error.

**Comment by remoteconn-7891** [comment]
This is what I got after I ran nmp install:




coreyvigil@Coreys-Air virtualify % npm install
npm ERR! code EJSONPARSE
npm ERR! path /Applications/ProjectVirtual/virtualify/package.json
npm ERR! JSON.parse Bad control character in string literal in JSON at position 123 (line 4 column 80) while parsing '{
npm ERR! JSON.parse "name": "Virtualify",
npm ERR! JSON.parse "version": "2.0.'
npm ERR! JSON.parse Failed to parse JSON data.
npm ERR! JSON.parse Note: package.json must be actual JSON, not just JavaScript.

npm ERR! A complete log of this run can be found in: /Users/coreyvigil/.npm/_logs/2023-11-15T07_19_09_375Z-debug-0.log

**Reply by stolenfallen1** [reply]
Did you copy the codes I sent as a replacement to your package.json before running npm install?

**Comment by remoteconn-7891** [comment]
Yes, I did. I'll re-copy it

**Comment by remoteconn-7891** [comment]
I got work in the morning, wanna try this again tomorrow evening? I'm MST by the way

**Reply by stolenfallen1** [reply]
Sure, regarding your problem with bad control character string literal this StackOverflow discussion might help

https://stackoverflow.com/questions/17487058/how-to-solve-json-parse-bad-control-character-in-string-literal-in-this-code

**Reply by remoteconn-7891** [reply]
Yeah I was reading this before you messaged me. Hopefully we can get this resolved tomorrow. The code is correct, so I have no idea why it's doing this

**Reply by stolenfallen1** [reply]
Yeah sure no problem, I'll just leave it here for your references tomorrow.

After the progress we have so far, I think there are still underlying problems somewhere in your code in a different file, where you are parsing JSON type data. I think this GitHub discussion might help you with that.

https://github.com/langchain-ai/langchainjs/issues/2902

**Comment by remoteconn-7891** [comment]
I also have auth.js, user.js files on top of index.js files. I can copy/paste the code for all 3 files here

**Comment by remoteconn-7891** [comment]
Keep in mind, most of this code was from that tutorial video. So here is my auth.js

`const express = require("express");
const cors = require("cors");
//Using bodyParser as a Middleware function
const bodyParser = require('body-parser');
const bcryptjs = require("bcryptjs");
const authRouter = express.Router();



//Create an Account
authRouter.post("/ap/signup", async (req, res) => {
try {
    const { name, email, password, role } = req.body;
 
    const existingUser = await User.findOne({ email });
    if (existingUser) {
        return res
          .status(400)
          .json({ msg: "User with this email already exists." });
    }

    const hashedPassword = await bcryptjs.hash(password, 8);

    //User login credentials include user, email address, password. I also
    //included a role field to differentiate between remote job seeker and
    //company
  let user = new User({
    email,
    password: hashedPassword,
    name,
    role,
  });
  user = await user.save();
  res.json(user);
} catch (e) {
  res.status(500).json({ error: e.message });
}
});

// Example login route
authRouter.post("/login", async (req, res) => {
    const { email, password } = req.body;

    try {
        const user = await User.findOne({ email, password });
        //This 401 error code just insinuates that it failed to
        //authenticate the user's login credentials
        if (!user) {
            return res.status(401).json({ message: "Invalid credentials" });
        }

        // Check the user's role and redirect accordingly
        if (user.role === "job_seeker") {
            // Redirect to job seeker dashboard
            res.redirect("/job_seeker/dashboard");
        } else if (user.role === "company") {
            // Redirect to company dashboard
            res.redirect("/company/dashboard");
        } else {
            res.status(401).json({ message: "Invalid role" });
        }
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: "Internal server error" });
    }
});

module.exports = authRouter;
`

**Comment by remoteconn-7891** [comment]
Here is the code for user.js:


`const mongoose = require("mongoose");

const userSchema = mongoose.Schema({
    name: {
        required: true,
        type: String,
        trim: true,
    },
    email: {
        required: true,
        type: String,
        trim: true,
        unique: true,
        validate: {
            validator: (value) => {
                // Regular expression for a simple email validation
                const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
                return re.test(value.toLowerCase()); // Convert to lowercase to ensure case-insensitive validation
            },
            message: "Invalid email format",
        },
    },
    password: {
        required: true,
        type: String,
        validate: {
            validator: (value) => {
                return value.length >= 8; // Minimum password length of 8 characters
            },
            message: "Password must be at least 8 characters long",
        },
    },
    role: {
        required: true,
        type: String,
        enum: ["job_seeker", "company"], // Enum to define possible roles
    },
});

const User = mongoose.model("User", userSchema);

module.exports = User;`

**Comment by remoteconn-7891** [comment]
I already sent it, but here is index.js. I merely replaced const with var, don't know if that makes a difference or not:

`var express = require("express");
var mongoose = require("mongoose");
var dotenv = require("dotenv"); // Import dotenv 
var authRouter = require("./routes/auth");
dotenv.config(); // Load environment variables // Load environment variables from .env file
var PORT = process.env.PORT || 3000;
var app = express();


app.use(json());
app.use(authRouter);
//DB just stands for MongoDB database, this String is the password for my database
const DB = 
"mongodb+srv://coreyjames689:remotify87@cluster0.jybj21u.mongodb.net/Cluster0?retryWrites=true&w=majority";

connect(DB)
  .then(() => {
    console.log("Connection Successful");
  })
  // catch(e) here is used for error handling by catching coding errors in Express
  .catch((e) => {
    console.log(e);
  });


  // My server is listening on PORT 3000
app.listen(3000, () => {
  console.log('Server is running on port ${(3000)}');
});`

**Comment by remoteconn-7891** [comment]
These are all the files associated with node.jS

**Comment by remoteconn-7891** [comment]
Now that you can see the code for all the express files, do you see any errors?

**Comment by rangav** [comment]
Try this solutions 
https://github.com/rangav/thunder-client-support/blob/master/docs/common-issues.md
